### ISSUE TITLE
[Question] Is there anything we can do about faster support of kubernetes versions?

### ISSUE THREAD

**Issue by ItsReddi** [rancher/rke#3441]
Hi,
since kubernetes versions support 1.27 & 1.28 are still on the road.
- 1.27 maybe supported "soon" https://github.com/rancher/rancher/issues/41840
- 1.28 supported at sometime next year? https://github.com/rancher/rancher/issues/43109
- 1.29 released in a few days (could not find a issue)

The issue is the short time between Kubernetes EOL and Rancher supporting it.
Which is somehow forcing us to rush new rancher update sometimes, even if there are known issues that would be fixed in the next rancher release that mostly have nothing todo with the kubernetes version itself.

Current example:
1.26 => 1.27 => 1.28
1.26 EOL (in 12 weeks)
1.27 EOL (7 Month)
1.28 EOL (11 Month)

Current upgrade plan:
- We hope for a 1.27 supporting release in the next two weeks.
- We hope some issues discovered and fixed early (but Christmas holidays will slow that down)
- Upgrade target end of January (EOL left 1.27: 5 Month)
- Wait for 1.28 rancher support, hopefully May, EOL 1.27 (28.06.24)
- and so on, waiting for rushing 1.29

That are many hopes :)

Is there anything we can do about supporting versions faster?

**Comment by superseb** [comment]
@leflambeur can you please answer this

**Comment by leflambeur** [comment]
Absolutely @superseb 

Currently support for a new Kubernetes Minor version in Rancher Manager is done in a 'big-bang'.

This means we align support for not just RKE1, RKE2 and K3s, but also AKS, EKS, GKE in a Rancher Release (some hardcoding requirements in the codebase due to restrictions on upstream APIs etc) and also have to schedule this across our development/code-freeze dates.

This is a practice we have followed since pretty much the beginning and was largely manageable - in more recent times though we are finding aligning these different lifecycles has become much more difficult especially with some recent changes upstream significantly impacting our other features (PSPs -> PSAs in v1.25, originally un-release-noted In-tree Provider Removals in v1.27).

For legacy reasons RKE1 is heavily tied to the Rancher lifecycle which typically means support for new RKE1 versions happens as Rancher itself supports them.

I am personally very keen to see support of new Kubernetes versions not just in RKE but Rancher Manager across all distributions ASAP, and have been advocating for this internally as much as possible and driving hard on this since I re-joined the team last year.

This could mean untangling the RKE lifecycle from Rancher but also investigating how we can maybe phase in support of a new minor for distributions vs the 'big-bang' approach we have traditionally followed.

Myself in product, and the engineering team are putting a lot of effort into catching up and speeding this process up properly with new dedicated resources on supporting new minors and beginning to re-architect certain dependencies from being embedded in our codebase, like CAPI for RKE2 Provisioning which has it's own Minor support lifecycle, to improve the overall speed of delivery (the CAPI change was introduced in v2.7.7+).

I originally targeted v1.28 as well as v1.27 for Rancher v2.8.0 in November but unfortunately this just missed our code-freeze as v1.27 introduced some friction on the deprecation of in-tree AWS provider support (which we discovered separately, and was not in the upstream release notes originally).

I have given our Engineering Team the challenge of delivering both v1.28 and v1.29 in our next feature release (Late Q1 24) to catch up ASAP but the primary target is at least v1.28 which I believe is already undergoing validation now as v1.29 is due upstream on 5th December and not yet released.

TL;DR - To support a new K8s version in Rancher Manager today requires aligning multiple other distributions lifecycles and timing that with our other committed dates. RKE1 is tied heavily to Rancher Manager's own support and usually is delivered simultaneously. We have been making changes in our architecture, codebase, and the resourcing we have to speed up how we deliver these requirements to our users ASAP. V1.27 support is targeted for v2.8 due Late Nov. v1.28 and v1.29 support have both been targeted for our Q1 24 release to catch up asap with v1.28 already undergoing validation, v1.29 is yet to be released upstream so is more at risk as a result.

**Comment by ItsReddi** [comment]
Thank you for the detailed answer on Rancher Manager's approach to Kubernetes support. It's great to hear about the efforts to speed up the process and the steps being taken to improve the system, especially in light of the challenges you've described. Your dedication to advocating for these changes is commendable, let's hope together than 👍.  Thank you.
